skcosmo/feature_selection/voronoi_fps.py

```diff
@@ -0,0 +1,136 @@
+
+    def _get_active(self, X, last_selected):
+        # because we haven't called super() yet, self.n_selected_ has not been
+        # incremented yet
+
+        f_active = np.zeros(self.n_features_to_select, bool)
+
+        """must compute distance of the new point to all the previous FPS. Some
+            of these might have been computed already, but bookkeeping could be
+            worse that recomputing (TODO: verify!)
+            """
+        # calculation in a single block
+        self.sel_d2q_[: self.n_selected_] = (
+            self.norms_[self.selected_idx_[: self.n_selected_]]
+            + self.norms_[last_selected]
+            - 2 * (self.X_selected_[:, : self.n_selected_].T @ X[:, last_selected])
+        ) * 0.25
+
+        for ic in range(self.n_selected_):
+            # empty voronoi, no need to consider it
+            if self.number_in_voronoi[ic] > 1:
+                r2 = max(self.haussdorf_[self.idx_in_voronoi[ic]])
+                if self.sel_d2q_[ic] < r2:
+                    # these voronoi cells need to be updated
+                    f_active[ic] = True
+        f_active[self.n_selected_] = True
+        return f_active
+
+    def _move_point(self, point, new_voronoi_idx):
+        old_voronoi_loc = self.voronoi_location[point]
+        self.voronoi_location[point] = new_voronoi_idx
+
+        self.number_in_voronoi[old_voronoi_loc] -= 1
+        self.idx_in_voronoi[old_voronoi_loc] = np.array(
+            np.setdiff1d(self.idx_in_voronoi[old_voronoi_loc], [point]), dtype=int
+        )
+        self.number_in_voronoi[new_voronoi_idx] += 1
+
+        self.idx_in_voronoi[new_voronoi_idx] = np.array(
+            np.concatenate((self.idx_in_voronoi[new_voronoi_idx], [point])), dtype=int
+        )
+
+    def _repartition(self, X, old_voronoi_idx, new_voronoi_idx):
+
+        points_in_voronoi = self.idx_in_voronoi[old_voronoi_idx]
+        true_idx = self.selected_idx_[new_voronoi_idx]
+
+        new_dist = (
+            self.norms_[points_in_voronoi]
+            + self.norms_[true_idx]
+            - 2 * X[:, points_in_voronoi].T @ X[:, true_idx]
+        )
+
+        updated_points = np.where(new_dist < self.haussdorf_[points_in_voronoi])[0]
+        l_update = points_in_voronoi[updated_points]
+
+        self.haussdorf_[l_update] = new_dist[updated_points]
+
+        for p in l_update:
+            self._move_point(p, new_voronoi_idx)
+
+    def _get_dist(self, X, last_selected):
+        f_active = self._get_active(X, last_selected)
+        self._move_point(last_selected, self.n_selected_)
+        self.haussdorf_[last_selected] = 0
+        if (
+            np.sum(self.number_in_voronoi[f_active]) / X.shape[1]
+            > self.voronoi_cutoff_fraction
+        ):
+            all_dist = super()._get_dist(X, last_selected)
+
+            updated_points = np.where(all_dist < self.haussdorf_)[0]
+            for p in updated_points:
+                if self.voronoi_location[p] != last_selected:
```

Comment:
I don't know, is this checking necessary? Why we can't move point  into his Voronoi polyhedron one another time? :) Maybe, it will be slightly faster than making a comparison every time.